### ISSUE TITLE
Android ofVideoPlayer Not Working with ofGLProgrammableRenderer

### ISSUE THREAD

**Issue by danoli3** [openframeworks/openFrameworks#3000]
When running a project with ofGLProgrammableRenderer the ofVideoPlayer in Android draws only a white area and not the video. (Even though video is loaded from disk).

```
#include "ofMain.h"
#include "ofApp.h"

int main(){
    
    ofSetCurrentRenderer(ofGLProgrammableRenderer::TYPE); // <--- Change
    
	ofSetupOpenGL(1024,768, OF_WINDOW);			// <-------- setup the GL context

	// this kicks off the running of my app
	// can be OF_WINDOW or OF_FULLSCREEN
	// pass in width and height too:
	ofRunApp( new ofApp() );
	return 0;
}


#ifdef TARGET_ANDROID
#include <jni.h>

//========================================================================
extern "C"{
	void Java_cc_openframeworks_OFAndroid_init( JNIEnv*  env, jobject  thiz ){
		main();
	}
}
#endif
```

LogCat Trace:
```
05-30 12:17:04.637: W/Adreno-ES20(14871): <process_gl_state_enables:511>: GL_INVALID_ENUM
05-30 12:17:04.637: E/MediaPlayer(14871): Should have subtitle controller already set
05-30 12:17:04.637: E/libEGL(14871): called unimplemented OpenGL ES API
05-30 12:17:04.958: V/ofxAndroidVideoPlayer(14871): loadMovie(): movie size: 720x576
05-30 12:17:05.118: W/MediaPlayer(14871): info/warning (3, 0)
```

Video plays normally and correctly without ofGLProgrammableRenderer.
Normal LogCat Trace in project without ofGLProgrammableRenderer set:
```
05-30 13:18:54.546: E/MediaPlayer(17547): Should have subtitle controller already set
05-30 13:18:55.057: W/MediaPlayer(17547): info/warning (3, 0)
05-30 13:18:55.067: W/GLConsumer(17547): [unnamed-17547-0] bindTextureImage: clearing GL error: 0x500
```

Will follow up with test project

**Comment by davivid** [comment]
Hi,
Did you fix this by any chance?

**Comment by geoffdonaldson** [comment]
@arturoc Can you point me in the right direction on how to fix this. I really need this feature. I've modified the ofGLProgrammableRenderer to include a shader that uses the samplerExternalOES sampler when the texture target is GL_TEXTURE_EXTERNAL_OES but am still having issues getting it to work.

**Comment by schmitzware** [comment]
Any follow up on this? I'm in the same boat and need to get video working on Android with ES 2+. I'm just starting to look into the notes above.

**Comment by schmitzware** [comment]
I'm now at the same point as Geoff but am still not getting anything other than a black rectangle.

Texture target is GL_TEXTURE_EXTERNAL_OES 
Sampler is samplerExternalOES
The movie plays fine (I can hear it).
I've stuffed in a bunch of glGetError calls and everything seem to be running fine but no texture. 

I have this running on OS X (same shader other than the sampler) and it's rendering perfectly.

Anyone have any ideas????

**Comment by schmitzware** [comment]
Also, the errors from the original post 

05-30 12:17:04.637: W/Adreno-ES20(14871): <process_gl_state_enables:511>: GL_INVALID_ENUM

are from the glEnable/glDisable calls in ofxAndroidVideoPlayer::reloadTexture().

FYI.

**Comment by schmitzware** [comment]
So, I have this working and thought I'd post the fix here for other folks and for future OpenFrameworks releases.  There were a couple things going on in OFAndroidVideoPlayer.java & ofxAndroidVideoPlayer.cpp. I'm including my versions below so you can diff them.

In a nutshell, the order of creating an Android movie player and assigning the texture surface to renderer to was not quite right (according to the docs in MediaPlayer). So, I've rearranged the order. The other VERY important piece is that the ofTexture in ofxAndroidVideoPlayer.cpp needs to have 'setUseExternalTextureID' called with the generated texture id, otherwise glDeleteTextures will get called down in the depths of OpenFrameworks who ruins all the fun.

I haven't tested reloading things with a different movie file but in theory it should work. Buyer beware.

Here's the 2 files

------------------------------- OFAndroidVideoPlayer.java ----------------------------------------

package cc.openframeworks;

import android.annotation.TargetApi;
import android.graphics.SurfaceTexture;
import android.graphics.SurfaceTexture.OnFrameAvailableListener;
import android.media.MediaPlayer;
import android.media.MediaPlayer.OnPreparedListener;
import android.media.MediaPlayer.OnVideoSizeChangedListener;
import android.media.MediaPlayer.OnCompletionListener;
import android.os.Build;
import android.util.FloatMath;
import android.util.Log;
import android.view.Surface;


@TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH)
public class OFAndroidVideoPlayer extends OFAndroidObject implements OnFrameAvailableListener {
	
	public OFAndroidVideoPlayer(){
		
		bIsLoaded = false;
		bIsPlaying = false;
		bIsPaused = true;
		bIsFrameNew = false;
		bAutoResume = false;
		bIsMoviedone = false;
		
		// TODO Get movie FPS to implement Frame methods
		// movieFPS = 16;
		// nFrames = 0;
		
		pan = 0.f;
		volume = leftVolume = rightVolume = 1;

		mediaPlayer = new MediaPlayer();
		mediaPlayer.setOnPreparedListener(new OnPreparedListener() {
				public void onPrepared(MediaPlayer mp) {
					bIsLoaded = true;
					if(bAutoResume) {
						setPositionMS(movieResumeTime);
						bAutoResume = false;
						play();
					}
				}
			});
			mediaPlayer.setOnVideoSizeChangedListener(new OnVideoSizeChangedListener() {
				public void onVideoSizeChanged(MediaPlayer mp, int width,
																			 int height) {
					// TODO Auto-generated method stub
				}
			});
			mediaPlayer.setOnCompletionListener(new OnCompletionListener() {
				public void onCompletion(MediaPlayer mp) {
					bIsMoviedone = true;
				}
			});
	}
	
	public void setTexture(int texName) {
		surfaceTexture = new SurfaceTexture(texName);
		surfaceTexture.setOnFrameAvailableListener(this);
		surface = new Surface(surfaceTexture);
		mediaPlayer.setSurface(surface);
	}
	
	public void clearTextures() {
		if(surface != null) {
			surface.release();
			surface = null;
		}
		if(surfaceTexture != null) {
			surfaceTexture.setOnFrameAvailableListener(null);
			surfaceTexture.release();
			surfaceTexture = null;
		}
	}
	
	public boolean update() {
		synchronized(this){
			if(bIsFrameNew) {
				if(surfaceTexture != null)
					surfaceTexture.updateTexImage();
				bIsFrameNew = false;
				bIsMoviedone = false;
				return true;
			} else {
				return false;
			}
		}
	}
	
	public void getTextureMatrix(float[] mtx) {
		if(surfaceTexture != null) surfaceTexture.getTransformMatrix(mtx);
	}
	
	public void loadMovie(String fileName){
		try {
			this.fileName = fileName;
			//mediaPlayer.reset();
			mediaPlayer.setDataSource(fileName);
			mediaPlayer.prepare();
		} catch (Exception e) {
			Log.e("OF","couldn't load " + fileName,e);
		} 		
	}
	
	public void play(){

		if(mediaPlayer==null) return;
		
		if(!bIsLoaded){
			Log.e("OF","ofxAndroidVideo::play - movie not loaded!");
			return;
		}
		
		mediaPlayer.start();
		
		bIsPlaying = true;
		bIsPaused = false;
	}
	
	public void stop(){
		if(mediaPlayer==null) return;
		mediaPlayer.stop();
		bIsPlaying = false;
	}
	
	void unloadMovie(){
		if(mediaPlayer!=null){
			mediaPlayer.setSurface(null);
			mediaPlayer.release();
			mediaPlayer = null;
		}
		clearTextures();
		
		fileName = null;
		bIsLoaded = false;
		bIsMoviedone = false;
		bIsPlaying = false;
		bIsPaused = true;
		
		//nFrames = 0;
		
	}
	
	void setVolume(float vol){
		volume = vol;
		// calculates left/right volumes from pan-value (constant panning law) 
		// see: Curtis Roads: Computer Music Tutorial p 460
		// thanks to jasch
		float angle = pan * 0.7853981633974483f; // in radians from -45. to +45.
		float cosAngle = FloatMath.cos(angle);
		float sinAngle = FloatMath.sin(angle);
		leftVolume  = (float)((cosAngle - sinAngle) * 0.7071067811865475) * vol; // multiplied by sqrt(2)/2
		rightVolume = (float)((cosAngle + sinAngle) * 0.7071067811865475) * vol; // multiplied by sqrt(2)/2
		if(mediaPlayer!=null)mediaPlayer.setVolume(leftVolume, rightVolume);
	}
	
	float getVolume(){
		return volume;
	}
	
	void setPaused(boolean bP){
		if(mediaPlayer==null) return;
		if(bP) {
			if(bIsPlaying) {
				mediaPlayer.pause();
				bIsPlaying = !bP;
				bIsPaused = bP;
			}
		} else {
			if(bIsPaused) {
				mediaPlayer.start();
				bIsPlaying = !bP;
				bIsPaused = bP;
			}
		}
	}
	
	void setLoopState(boolean bL){
		if(mediaPlayer==null) return;
		mediaPlayer.setLooping(bL);
	}
	
	boolean getLoopState(){
		if(mediaPlayer==null) return false;
		return mediaPlayer.isLooping();
	}
	
	int getWidth(){
		if(mediaPlayer==null) return 0;
		return mediaPlayer.getVideoWidth();
	}
	
	int getHeight(){
		if(mediaPlayer==null) return 0;
		return mediaPlayer.getVideoHeight();
	}
	
	/* Needs movie frameRate to work
	int	getCurrentFrame(){
		
		float  framePosInFloat = ((float)getTotalNumFrames() * (float)getPosition());
		int    framePosInInt = (int)framePosInFloat;
		float  floatRemainder = (framePosInFloat - framePosInInt);
		if (floatRemainder > 0.5) framePosInInt = framePosInInt + 1;

		return framePosInInt;

	}
	*/
	
	/* Needs movie frameRate to work
	int	getTotalNumFrames(){
		return nFrames;
	}
	*/
		
	public boolean isLoaded(){
		return bIsLoaded;
	}
	
	public boolean isPaused(){
		return bIsPaused;
	}
	
	public boolean isPlaying(){
		return bIsPlaying;
	}
	
	boolean isMovieDone(){
		return bIsMoviedone;
	}
	
	
	void setPosition(float pct){
		if(mediaPlayer!=null) mediaPlayer.seekTo((int) (mediaPlayer.getDuration()*pct)); // 0 = start, 1 = end;
	}
	
	void setPositionMS(int ms){
		if(mediaPlayer!=null) mediaPlayer.seekTo(ms); // 0 = start, 1 = end;
	}
	
	float getPosition(){
		if(mediaPlayer!=null)
			return ((float)mediaPlayer.getCurrentPosition())/(float)mediaPlayer.getDuration();
		else
			return 0;
	}
	
	int getPositionMS(){
		if(mediaPlayer!=null)
			return mediaPlayer.getCurrentPosition();
		else
			return 0;
	}
	
	float getDuration(){
		if(mediaPlayer!=null)
			return (float)mediaPlayer.getDuration();
		else
			return 0;
	}
	
	int getDurationMS(){
		if(mediaPlayer!=null)
			return mediaPlayer.getDuration();
		else
			return 0;
	}
	

	
	@Override
	protected void appPause() {
		int currMovieTime = getPositionMS();
		stop();
		String currFileName = fileName;
		boolean currIsLoaded = bIsLoaded; 
		boolean currIsPlaying = bIsPlaying;
		unloadMovie();
		fileName = currFileName;
		bIsLoaded = currIsLoaded;	
		bIsPlaying = currIsPlaying;
		
		bAutoResume = true;
		movieResumeTime = currMovieTime;
	
	}

	@Override
	protected void appResume() {
		if(bIsLoaded){
			loadMovie(fileName);
		}
	}

	@Override
	protected void appStop() {
		appPause();
	}
	
	@Override
	public void onFrameAvailable(SurfaceTexture arg0) {
		synchronized(this){
			bIsFrameNew = true;
		}
	}
	
	
	
	private MediaPlayer mediaPlayer;
	
	private SurfaceTexture surfaceTexture;
	private Surface surface;
	private String fileName;
	private float pan;
	private float volume;
	private float leftVolume, rightVolume;
	private boolean bIsLoaded;
	private boolean bIsPlaying;
	private boolean bIsPaused;
	private boolean bIsMoviedone;
	private boolean bIsFrameNew;
	
	//private int movieFPS;
	//private int nFrames;
	
	private boolean bAutoResume;
	private int movieResumeTime;
	
	
}
 
------------------------------- ofxAndroidVideoPlayer.cpp ----------------------------------------

/*
 * ofxAndroidVideoPlayer.cpp
 *
 *  Created on: 25/04/2013
 *      Author: p
 */

#include "ofxAndroidVideoPlayer.h"
#include "ofxAndroidUtils.h"
#include "ofLog.h"


//---------------------------------------------------------------------------
void ofxAndroidVideoPlayer::reloadTexture(){

	texture.texData.textureTarget 	= GL_TEXTURE_EXTERNAL_OES;
    texture.texData.glTypeInternal 	= GL_RGBA;
    texture.texData.bFlipTexture 	= false;

	glGenTextures(1, (GLuint *)&texture.texData.textureID);

	glEnable(texture.texData.textureTarget);

	glBindTexture(texture.texData.textureTarget, (GLuint)texture.texData.textureID);

	glTexParameterf(texture.texData.textureTarget, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
	glTexParameterf(texture.texData.textureTarget, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
	glTexParameterf(texture.texData.textureTarget, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
	glTexParameterf(texture.texData.textureTarget, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
	glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_MODULATE);

	glDisable(texture.texData.textureTarget);

	// This is very important - otherwise the glDeleteTextures is called
	texture.setUseExternalTextureID(texture.texData.textureID);

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "reloadTexture(): couldn't get environment using GetEnv()";
		return;
	}

	jmethodID javasetTexture = env->GetMethodID(javaClass,"setTexture","(I)V");
	if(!javasetTexture){
		ofLogError("ofxAndroidVideoPlayer") << "reloadTexture(): couldn't get java setTexture for VideoPlayer";
		return;
	}

  	env->CallVoidMethod(javaVideoPlayer,javasetTexture,texture.texData.textureID);
}

//---------------------------------------------------------------------------
void ofxAndroidVideoPlayer::unloadTexture(){
	texture.texData.textureID=0;
}


//---------------------------------------------------------------------------
ofxAndroidVideoPlayer::ofxAndroidVideoPlayer(){

	texture.texData.bAllocated = false;

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "constructor: couldn't get environment using GetEnv()";
		return;
	}
	jclass localClass = env->FindClass("cc/openframeworks/OFAndroidVideoPlayer");
	javaClass = (jclass) env->NewGlobalRef(localClass);

	if(!javaClass){
		ofLogError("ofxAndroidVideoPlayer") << "constructor: couldn't get java class for VideoPlayer";
		return;
	}

	jmethodID constructor = env->GetMethodID(javaClass,"<init>","()V");
	if(!constructor){
		ofLogError("ofxAndroidVideoPlayer") << "constructor: couldn't get java constructor for VideoPlayer";
		return;
	}

	javaVideoPlayer = env->NewObject(javaClass,constructor);
	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "constructor: couldn't create java VideoPlayer";
		return;
	}

	javaVideoPlayer = (jobject)env->NewGlobalRef(javaVideoPlayer);

	bIsFrameNew = false;

	jfloatArray localMatrixJava = env->NewFloatArray(16);
	matrixJava = (jfloatArray) env->NewGlobalRef(localMatrixJava);

	reloadTexture();

	ofAddListener(ofxAndroidEvents().unloadGL,this,&ofxAndroidVideoPlayer::unloadTexture);
	ofAddListener(ofxAndroidEvents().reloadGL,this,&ofxAndroidVideoPlayer::reloadTexture);
}

//---------------------------------------------------------------------------
ofxAndroidVideoPlayer::~ofxAndroidVideoPlayer(){
	JNIEnv *env = ofGetJNIEnv();
	if(javaVideoPlayer) env->DeleteGlobalRef(javaVideoPlayer);
	if(javaClass) env->DeleteGlobalRef(javaClass);
	if(matrixJava) env->DeleteGlobalRef(matrixJava);
}


//---------------------------------------------------------------------------
bool ofxAndroidVideoPlayer::load(string fileName){

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "load(): java soundVideoPlayer not loaded";
		return false;
	}

	JNIEnv *env = ofGetJNIEnv();
	jmethodID javaLoadMethod = env->GetMethodID(javaClass,"loadMovie","(Ljava/lang/String;)V");
	if(!javaLoadMethod){
		ofLogError("ofxAndroidVideoPlayer") << "load(): couldn't get java loadVideo for VideoPlayer";
		return false;
	}

	jstring javaFileName = ofGetJNIEnv()->NewStringUTF(ofToDataPath(fileName,true).c_str());
	env->CallVoidMethod(javaVideoPlayer,javaLoadMethod,javaFileName);
	env->DeleteLocalRef((jobject)javaFileName);

	jmethodID javaGetWidthMethod = env->GetMethodID(javaClass,"getWidth","()I");
	if(!javaGetWidthMethod){
		ofLogError("ofxAndroidVideoPlayer") << "loadMovie(): couldn't get java getWidth for VideoPlayer";
		return false;
	}
	width = env->CallIntMethod(javaVideoPlayer, javaGetWidthMethod);

	jmethodID javaGetHeightMethod = env->GetMethodID(javaClass,"getHeight","()I");
	if(!javaGetHeightMethod){
		ofLogError("ofxAndroidVideoPlayer") << "loadMovie(): couldn't get java getHeight for VideoPlayer";
		return false;
	}
	height = env->CallIntMethod(javaVideoPlayer, javaGetHeightMethod);

	// hack to initialize gl resources from outside ofTexture
	texture.texData.bAllocated	= true;
	texture.texData.width 		= width;
	texture.texData.height 		= height;
	texture.texData.tex_w 		= width;
	texture.texData.tex_h 		= height;
	texture.texData.tex_t 		= 1; // Hack!
	texture.texData.tex_u 		= 1;

	ofLogVerbose("ofxAndroidVideoPlayer") << "loadMovie(): movie size: "<< width << "x" << height << ":" << texture.texData.textureTarget;

	return true;
}

//---------------------------------------------------------------------------
void ofxAndroidVideoPlayer::close(){

}

//---------------------------------------------------------------------------
void ofxAndroidVideoPlayer::update(){

	JNIEnv *env = ofGetJNIEnv();
	jmethodID javaUpdate = env->GetMethodID(javaClass,"update","()Z");
	if(!javaUpdate){
		ofLogError("ofxAndroidVideoPlayer") << "update(): couldn't get java update for VideoPlayer";
		return;
	}
	bIsFrameNew = env->CallBooleanMethod(javaVideoPlayer,javaUpdate);

	jmethodID javaGetTextureMatrix = env->GetMethodID(javaClass,"getTextureMatrix","([F)V");
	if(!javaGetTextureMatrix){
		ofLogError("ofxAndroidVideoPlayer") << "update(): couldn't get java javaGetTextureMatrix for VideoPlayer";
		return;
	}
	env->CallVoidMethod(javaVideoPlayer,javaGetTextureMatrix,matrixJava);
	jfloat * m = env->GetFloatArrayElements(matrixJava,0);

	ofMatrix4x4 textureMatrix(m);

	ofMatrix4x4 vFlipTextureMatrix;
	vFlipTextureMatrix.scale(1,-1,1);
	vFlipTextureMatrix.translate(0,1,0);
	texture.setTextureMatrix(vFlipTextureMatrix * textureMatrix);
	//texture.getTextureData().tex_t = 1.+1-matrix.getPtr()[0]; // Hack!
	//texture.getTextureData().tex_u = 1.;

	env->ReleaseFloatArrayElements(matrixJava,m,0);

}

//---------------------------------------------------------------------------
void ofxAndroidVideoPlayer::play(){

	JNIEnv *env = ofGetJNIEnv();
	jmethodID javaPlay = env->GetMethodID(javaClass,"play","()V");
	if(!javaPlay){
		ofLogError("ofxAndroidVideoPlayer") << "play(): couldn't get java play for VideoPlayer";
		return;
	}
	env->CallVoidMethod(javaVideoPlayer,javaPlay);

};

//---------------------------------------------------------------------------
void ofxAndroidVideoPlayer::stop(){

	JNIEnv *env = ofGetJNIEnv();
	jmethodID javaStop = env->GetMethodID(javaClass,"stop","()V");
	if(!javaStop){
		ofLogError("ofxAndroidVideoPlayer") << "stop(): couldn't get java stop for VideoPlayer";
		return;
	}
	env->CallVoidMethod(javaVideoPlayer,javaStop);

};


//---------------------------------------------------------------------------
bool ofxAndroidVideoPlayer::isPaused() const {

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "isPaused(): java VideoPlayer not loaded";
		return 0;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "isPaused(): couldn't get environment using GetEnv()";
		return 0;
	}

	jmethodID javaIsPausedMethod = env->GetMethodID(javaClass,"isPaused","()Z");
	if(!javaIsPausedMethod){
		ofLogError("ofxAndroidVideoPlayer") << "isPaused(): couldn't get java isPaused for VideoPlayer";
		return 0;
	}

	return env->CallBooleanMethod(javaVideoPlayer,javaIsPausedMethod);

};


//---------------------------------------------------------------------------
bool ofxAndroidVideoPlayer::isLoaded() const {

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "isLoaded(): java VideoPlayer not loaded";
		return 0;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "isLoaded(): couldn't get environment using GetEnv()";
		return 0;
	}

	jmethodID javaIsLoadedMethod = env->GetMethodID(javaClass,"isLoaded","()Z");
	if(!javaIsLoadedMethod){
		ofLogError("ofxAndroidVideoPlayer") << "isLoaded(): couldn't get java isLoaded for VideoPlayer";
		return 0;
	}

	return env->CallBooleanMethod(javaVideoPlayer,javaIsLoadedMethod);

};

//---------------------------------------------------------------------------
bool ofxAndroidVideoPlayer::isPlaying() const {

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "isPlaying(): java VideoPlayer not loaded";
		return 0;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "isPlaying(): couldn't get environment using GetEnv()";
		return 0;
	}

	jmethodID javaIsPlayingMethod = env->GetMethodID(javaClass,"isPlaying","()Z");
	if(!javaIsPlayingMethod){
		ofLogError("ofxAndroidVideoPlayer") << "isPlaying(): couldn't get the java isPlaying for VideoPlayer";
		return 0;
	}

	return env->CallBooleanMethod(javaVideoPlayer,javaIsPlayingMethod);

};

//---------------------------------------------------------------------------
ofTexture * ofxAndroidVideoPlayer::getTexturePtr(){
	return & texture;
}

//---------------------------------------------------------------------------
float ofxAndroidVideoPlayer::getWidth() const {
	return width;
};

//---------------------------------------------------------------------------
float ofxAndroidVideoPlayer::getHeight() const {
	return height;
};

//---------------------------------------------------------------------------
float ofxAndroidVideoPlayer::getPosition() const {

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "getPosition(): java VideoPlayer not loaded";
		return 0.;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "getPosition(): couldn't get environment using GetEnv()";
		return 0;
	}

	jmethodID javaGetPositionMethod = env->GetMethodID(javaClass,"getPosition","()F");
	if(!javaGetPositionMethod){
		ofLogError("ofxAndroidVideoPlayer") << "getPosition(): couldn't get java getPosition for VideoPlayer";
		return 0;
	}

	return env->CallFloatMethod(javaVideoPlayer,javaGetPositionMethod);

};

//---------------------------------------------------------------------------
float ofxAndroidVideoPlayer::getDuration() const {

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "getDuration(): java VideoPlayer not loaded";
		return 0.;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "getDuration(): couldn't get environment using GetEnv()";
		return 0;
	}

	jmethodID javaGetDurationMethod = env->GetMethodID(javaClass,"getDuration","()F");
	if(!javaGetDurationMethod){
		ofLogError("ofxAndroidVideoPlayer") << "getDuration(): couldn't get java getDuration for VideoPlayer";
		return 0;
	}

	return env->CallFloatMethod(javaVideoPlayer,javaGetDurationMethod);

};

//---------------------------------------------------------------------------
bool ofxAndroidVideoPlayer::getIsMovieDone() const {

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "getIsMovieDone(): java VideoPlayer not loaded";
		return 0;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "getIsMovieDone(): couldn't get environment using GetEnv()";
		return 0;
	}

	jmethodID javaIsMovieDoneMethod = env->GetMethodID(javaClass,"isMovieDone","()Z");
	if(!javaIsMovieDoneMethod){
		ofLogError("ofxAndroidVideoPlayer") << "getIsMovieDone(): couldn't get java isMovieDone for VideoPlayer";
		return 0;
	}

	return env->CallBooleanMethod(javaVideoPlayer,javaIsMovieDoneMethod);

};

//---------------------------------------------------------------------------
void ofxAndroidVideoPlayer::setPosition(float pct){

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "setPosition(): java VideoPlayer not loaded";
		return;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "setPosition(): couldn't get environment using GetEnv()";
		return;
	}

	jmethodID javaSetPositionMethod = env->GetMethodID(javaClass,"setPosition","(F)V");
	if(!javaSetPositionMethod){
		ofLogError("ofxAndroidVideoPlayer") << "setPosition(): couldn't get java setPosition for VideoPlayer";
		return;
	}

	env->CallVoidMethod(javaVideoPlayer,javaSetPositionMethod,pct);

};

//---------------------------------------------------------------------------
void ofxAndroidVideoPlayer::setPaused(bool bPause){

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "setPaused(): java VideoPlayer not loaded";
		return;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "setPaused(): couldn't get environment using GetEnv()";
		return;
	}

	jmethodID javaPausedMethod = env->GetMethodID(javaClass,"setPaused","(Z)V");
	if(!javaPausedMethod){
		ofLogError("ofxAndroidVideoPlayer") << "setPaused(): coulnd't get java setPaused for VideoPlayer";
		return;
	}

	env->CallVoidMethod(javaVideoPlayer,javaPausedMethod,bPause?1:0);

};

//------------------------------------------------------------
void ofxAndroidVideoPlayer::setVolume(float vol){

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "setVolume(): java VideoPlayer not loaded";
		return;
	}
	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "setVolume(): couldn't get environment using GetEnv()";
		return;
	}

	jmethodID javaVolumeMethod = env->GetMethodID(javaClass,"setVolume","(F)V");
	if(!javaVolumeMethod){
		ofLogError("ofxAndroidVideoPlayer") << "setVolume(): couldn't get java setVolume for VideoPlayer";
		return;
	}

	env->CallVoidMethod(javaVideoPlayer,javaVolumeMethod,vol);

};

//------------------------------------------------------------
void ofxAndroidVideoPlayer::setLoopState(ofLoopType state){

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "setLoopState(): java VideoPlayer not loaded";
		return;
	}
	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "setLoopState(): couldn't get environment using GetEnv()";
		return;
	}

	jmethodID javaLoopStateMethod = env->GetMethodID(javaClass,"setLoopState","(Z)V");
	if(!javaLoopStateMethod){
		ofLogError("ofxAndroidVideoPlayer") << "setLoopState(): couldn't get java LoopState for VideoPlayer";
		return;
	}

	switch (state) {
		case OF_LOOP_NORMAL:
			env->CallVoidMethod(javaVideoPlayer,javaLoopStateMethod,1);
			break;
		case OF_LOOP_PALINDROME:
			// TODO Palindrome loop not implemented
			ofLogError("ofxAndroidVideoPlayer") << "setLoopState(): OF_LOOP_PALINDROME not implemented";
			break;
		case OF_LOOP_NONE:
			default:
			env->CallVoidMethod(javaVideoPlayer,javaLoopStateMethod,0);
			break;
	}

};

//------------------------------------------------------------
ofLoopType ofxAndroidVideoPlayer::getLoopState() const {

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "getLoopState(): java VideoPlayer not loaded";
		return OF_LOOP_NONE;
	}
	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "getLoopState(): couldn't get environment using GetEnv()";
		return OF_LOOP_NONE;
	}

	jmethodID javaGetLoopStateMethod = env->GetMethodID(javaClass,"getLoopState","()Z");
	if(!javaGetLoopStateMethod){
		ofLogError("ofxAndroidVideoPlayer") << "getLoopState(): couldn't get java GetLoopState for VideoPlayer";
		return OF_LOOP_NONE;
	}

	bool loopState = env->CallBooleanMethod(javaVideoPlayer,javaGetLoopStateMethod);

	if (loopState) {
		return OF_LOOP_NORMAL;
	} else {
		return OF_LOOP_NONE;
	}

};


//------------------------------------------------------------
/* Needs movie frameRate to work
int ofxAndroidVideoPlayer::getCurrentFrame() {

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "getCurrentFrame(): java VideoPlayer not loaded";
		return 0.;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "getCurrentFrame(): couldn't get environment using GetEnv()";
		return 0;
	}

	jmethodID javaGetCurrentFrame = env->GetMethodID(javaClass,"getCurrentFrame","()I");
	if(!javaGetCurrentFrame){
		ofLogError("ofxAndroidVideoPlayer") << "getCurrentFrame(): couldn't get java getCurrentFrame for VideoPlayer";
		return 0;
	}

	return env->CallIntMethod(javaVideoPlayer,javaGetCurrentFrame);

};
*/


//------------------------------------------------------------
/* Needs movie frameRate to work
int ofxAndroidVideoPlayer::getTotalNumFrames() {

	if(!javaVideoPlayer){
		ofLogError("ofxAndroidVideoPlayer") << "getTotalNumFrames(): java VideoPlayer not loaded";
		return 0.;
	}

	JNIEnv *env = ofGetJNIEnv();
	if (!env) {
		ofLogError("ofxAndroidVideoPlayer") << "getTotalNumFrames(): couldn't get environment using GetEnv()";
		return 0;
	}

	jmethodID javaGetTotalNumFrames = env->GetMethodID(javaClass,"getTotalNumFrames","()I");
	if(!javaGetTotalNumFrames){
		ofLogError("ofxAndroidVideoPlayer") << "getTotalNumFrames(): couldn't get java getTotalNumFrames for VideoPlayer";
		return 0;
	}

	return env->CallIntMethod(javaVideoPlayer,javaGetTotalNumFrames);

};
*/

**Comment by HalfdanJ** [comment]
@arturoc i'm looking at implementing the texture target correctly in the programmable renderer. I can see how to do it in the core, but it's a bit of a mess with android specific code in there. Where you thinking of putting it in there, or create some new class subclass of ofTexture to handle it in ofxAndroid?

**Comment by arturoc** [comment]
i think all we need to do is add a shader with the correct texture type that will be bound when a video player is bound and we are on android, we can also just check the texture target when binding a texture and if it's GL_EXTERNAL_OES bind the correct shader

**Comment by kkirby** [comment]
Just a heads up, there is also a very, very similar issue with the android camera on OpenGL ES 2+. The resulting image is white and there is a `W/GLConsumer(10076): [unnamed-10076-0] bindTextureImage: clearing GL error: 0x500` logged.

**Comment by HalfdanJ** [comment]
@kkirby this issue is fixed with #4590

**Comment by kkirby** [comment]
@HalfdanJ Thanks, worked perfectly!
